rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+- relying on existing oneDNN primitive attribute. This allows to
+  simplify documentation and usability by aligning semantics between
+  primitive and block level APIs. However, there is a set of
+  attributes that are specific to primitive API, and we might expose a
+  set of attributes specific to the block-level API (e.g. prefetching
+  strategy).
+- relying on dedicated attributes for block-level API. This allows a
+  clear separation of attributes between block level and primitive
+  level APIs, and allows to expose only supported attributes for block
+  level API.
+
+
+To make a first version of the API publicly available faster,
+we will start by relying on existing dnnl::primitive_attr objects.
+
+In the future, we will consider exposing a dedicated attribute for
+block-level API, to make API cleaner. Internally, we can rely on the
+same abstraction for primitive and block-level attributes to simplify
+passing attributes from primitive API calls to internal block level
+API calls. Like for primitive level API, those would be opaque, and
+attributes would be set/get using dedicated setters and getters.
+
+
+## Managing jitting overheads and caching support
+
+To implement a single matrix multiplication operation, the block-level
+API user will typically need to rely on multiple kernels.
+In particular:
+- tail handling on M, N and both M/N can require dedicated kernels if
+  M/N are not multiples of the block-size chosen by the user,
+- when blocking on K dimension, extra kernels are required to handle
+  first K-block (beta=0), intermediate K-blocks (beta=1) and last
+  K-block (beta=1, D matrix set, and post-ops specified).
+- when full matrices dimension change frequently. This will lead to
+  changing leading dimension at the brgemm level, requiring new
+  kernels.
+
+On top of that, if the block-level kernels are generated independantly
+for each layer of a neural network, the same kernels might be
+generated multiple times.
+
+Some API design decisions could impact potential reusability of kernels.
+In particular:
+- passing alpha/beta at runtime instead of creation time.
+- support two execute functions (with and without post-op and final
+  data conversion) for the same brgemm object.
+- support runtime leading dimensions.
+
+The proposal here is to stick with what is already tried and used
+internally. If during the experimental stage of the API, users request
+these functionalities, we will revisit the API and add the necessary
+support in the internal implementation. So alpha/beta will be creation
+time parameters, and kernels with and without post-ops will be handled
+with separate objects.  For runtime leading dimension, the internal
+API supports it but this will not be exposed for now. We can either
+change the API when we introduce the feature, or add a placeholder
+parameter if we are confident this feature will gain traction soon.
+
+Also, note that currently, no kernel-level cache exists for block
```

Comment:
The caching support is a must for performance. Frameworks have to provide their own cache support if oneDNN doesn't support it. oneDNN supports it at the beginning would be ideal.